src/mlpack/methods/linear_svm/linear_svm_main.cpp

```diff
@@ -0,0 +1,440 @@
+    }
+    if (CLI::HasParam("epochs"))
+    {
+      Log::Warn << PRINT_PARAM_STRING("epochs") << " ignored because "
+          << "optimizer type is not 'psgd'." << std::endl;
+    }
+  }
+
+  if (optimizerType != "lbfgs")
+  {
+    if (CLI::HasParam("max_iterations"))
+    {
+      Log::Warn << PRINT_PARAM_STRING("max_iterations") << " ignored because "
+          << "optimizer type is not 'lbfgs'." << std::endl;
+    }
+  }
+
+  // Step Size must be positive.
+  RequireParamValue<double>("step_size", [](double x) { return x >= 0.0; },
+      true, "step size must be positive");
+
+  // Lambda must be positive.
+  RequireParamValue<double>("lambda", [](double x) { return x >= 0.0; },
+      true, "lambda must be non-negative");
+
+  // Number of Classes must be Non-Negative
+  RequireParamValue<int>("num_classes", [](int x) { return x >= 0; },
+                         true, "number of classes must be greater than or "
+                         "equal to 0 (equal to 0 in case of unspecified.)");
+
+  // Delta must be positive.
+  RequireParamValue<double>("delta", [](double x) { return x >= 0.0; }, true,
+      "delta must be non-negative");
+
+  // Delta must be positive.
+  RequireParamValue<int>("epochs", [](int x) { return x > 0; }, true,
+      "epochs must be non-negative");
+
+  // These are the matrices we might use.
+  arma::mat trainingSet;
+  arma::Row<size_t> labels;
+  arma::Row<size_t> rawLabels;
+  arma::mat testSet;
+  arma::Row<size_t> predictedLabels;
+  size_t numClasses;
+
+  // Load data matrix.
+  if (CLI::HasParam("training"))
+    trainingSet = std::move(CLI::GetParam<arma::mat>("training"));
+
+  // Check if the labels are in a separate file.
+  if (CLI::HasParam("training") && CLI::HasParam("labels"))
+  {
+    rawLabels = std::move(CLI::GetParam<arma::Row<size_t>>("labels"));
+    if (trainingSet.n_cols != rawLabels.n_cols)
+    {
+      Log::Fatal << "The labels must have the same number of points as the "
+          << "training dataset." << endl;
+    }
+  }
+  else if (CLI::HasParam("training"))
+  {
+    // Checking the size of training data if no labels are passed.
+    if (trainingSet.n_rows < 2)
+    {
+      Log::Fatal << "Can't get labels from training data since it has less "
+          << "than 2 rows." << endl;
+    }
+
+    // The initial predictors for y, Nx1.
+    rawLabels = arma::conv_to<arma::Row<size_t>>::from(
+        trainingSet.row(trainingSet.n_rows - 1));
+    trainingSet.shed_row(trainingSet.n_rows - 1);
+  }
+
+  // Load the model, if necessary.
+  LinearSVMModel* model;
+  if (CLI::HasParam("input_model"))
+  {
+    model = CLI::GetParam<LinearSVMModel*>("input_model");
+  }
+  else
+  {
+    model = new LinearSVMModel();
+  }
+
+  // Now, do the training.
+  if (CLI::HasParam("training"))
+  {
+    data::NormalizeLabels(rawLabels, labels, model->mappings);
+    numClasses = CLI::GetParam<int>("num_classes") == 0 ?
+        arma::max(labels) + 1 : CLI::GetParam<int>("num_classes");
```

Comment:
We can avoid the `arma::max(labels)` by using the mappings we just produced: `model->mappings.n_elem` should be equivalent.

Reply:
Updated.